src/annotator/anchoring/trim-range.ts

```diff
@@ -0,0 +1,225 @@
+/**
+ * From which direction to evaluate strings or nodes: from the start of a string
+ * or range seeking Forwards, or from the end seeking Backwards.
+ */
+enum TrimDirection {
+  Forwards = 1,
+  Backwards,
+}
+
+/**
+ * An object representing metadata for a Range position (e.g. for use with
+ * Range.setStart or Range.setEnd)
+ */
+type RangePosition = {
+  offset: number;
+  node: Node;
+};
+
+/**
+ * Return the offset of the nearest non-whitespace character to `baseOffset`
+ * within `text`, looking in the `direction` indicated. Return -1 if no
+ * non-whitespace character exists between `baseOffset` and the terminus of the
+ * string (start or end depending on `direction`).
+ */
+function closestNonSpaceOffset(
+  text: string,
+  baseOffset: number,
+  direction: TrimDirection
+): number {
+  const nextChar =
+    direction === TrimDirection.Forwards ? baseOffset : baseOffset - 1;
+  if (text.charAt(nextChar).trim() !== '') {
+    // baseOffset is already a valid offset
+    return baseOffset;
+  }
+
+  let availableChars: string;
+  let availableNonWhitespaceChars: string;
+
+  if (direction === TrimDirection.Backwards) {
+    availableChars = text.substring(0, baseOffset);
+    availableNonWhitespaceChars = availableChars.trimEnd();
+  } else {
+    availableChars = text.substring(baseOffset);
+    availableNonWhitespaceChars = availableChars.trimStart();
+  }
+
+  if (!availableNonWhitespaceChars.length) {
+    return -1;
+  }
+
+  const offsetDelta =
+    availableChars.length - availableNonWhitespaceChars.length;
+
+  return direction === TrimDirection.Backwards
+    ? baseOffset - offsetDelta
+    : baseOffset + offsetDelta;
+}
+
+/**
+ * Calculate a new Range start position (TrimDirection.Forwards) or end position
+ * (Backwards) for `range` that represents the nearest non-whitespace character,
+ * moving into the `range` away from the relevant initial boundary node towards
+ * the terminating boundary node.
+ *
+ * The search is non-inclusive of the initial boundary node, but does consider
+ * the terminating boundary node.
+ *
+ * @throws {RangeError} If no text node with non-whitespace characters found
+ */
+function closestNonSpacePosition(
```

Comment:
This function returns _position_ metadata, ergo the choice of `closestNonSpacePosition` for its naming. Thoughts?

Reply:
This'll do. An alternative idea might be to name the functions based on what they search in, since that determines the output - eg. `closestNonSpaceInRange` and `closestNonSpaceInString`?

Reply:
I am using these latest suggestions, thanks!